src/box/memtx_space.c

```diff
@@ -1037,6 +1096,29 @@ memtx_build_on_replace(struct trigger *trigger, void *event)
 		if (stmt->old_tuple != NULL)
 			tuple_unref(stmt->old_tuple);
 	}
+	/*
+	 * Set on_rollback trigger on stmt to avoid
+	 * problem when rollbacked changes appears in
+	 * built-in-background index.
+	 */
+	struct on_rollback_trigger_with_data *on_rollback_associates =
+		region_aligned_alloc(&in_txn()->region,
+			sizeof(struct on_rollback_trigger_with_data),
+			alignof(struct on_rollback_trigger_with_data));
+	if (on_rollback_associates == NULL) {
+		diag_set(OutOfMemory,
+			 sizeof(struct on_rollback_trigger_with_data),
+			 "region_aligned_alloc",
+			 "memtx_build_on_replace");
+		diag_move(diag_get(), &state->diag);
+		return 0;
+	}
+	on_rollback_associates->data.stmt = stmt;
```

Comment:
Imho code looks a bit messy. Personally I would get rid of struct on_rollback_trigger_with_data and simply allocate single chunk of memory:
```
char *chunk = region_alloc(&in_txn()->region,
			sizeof(struct trigger) + sizeof(struct index_build_on_rollback_data));
...
struct trigger *on_rollback = chunk;
struct index_build_on_rollback_data *data = chunk + sizeof(*on_rollback);


```
Up to you